File: leetcode/python/328. 奇偶链表.py
#!/bin/bash python
"""
328. 奇偶链表
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: ListNode) -> ListNode:
        """
        新建两个头节点，分别用来存储奇偶节点
        """
        odd = curr_odd = ListNode()
        even = curr_even = ListNode()
        while head:
            # 存储奇数节点
            curr_odd.next = head
            curr_odd = curr_odd.next
            head = head.next
            # 存储偶数节点
            if head:
                curr_even.next = head
                curr_even = curr_even.next
                head = head.next
        curr_even.next = None
        curr_odd.next = even.next
        return odd.next

File: leetcode/cpp/328. 奇偶链表.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        ListNode *odd = new ListNode(), *curr_odd = odd;
        ListNode *even = new ListNode(), *curr_even = even;
        while (head) {
            curr_odd->next = head;
            curr_odd = curr_odd->next;
            head = head->next;
            if (head) {
                curr_even->next = head;
                curr_even = curr_even->next;
                head = head->next;
            }
        }
        curr_even->next = nullptr;
        curr_odd->next = even->next;
        return odd->next;
    }
};